File: src/ados/services/ui/dashboards/components/status_dot.py
"""Filled-circle status dot.

Used in the header (role indicator) and inside tiles (mesh / uplink
status). Color is the only signal — no shape variation for now since
we have three-color status semantics that map cleanly to
success/warning/error.
"""

from __future__ import annotations

from PIL import Image, ImageDraw

from . import primitives as p


def draw_dot(
    image: Image.Image,
    cx: int,
    cy: int,
    color: tuple[int, int, int],
    radius: int = 7,
) -> None:
    """Draw a filled circle centered on ``(cx, cy)``.

    Default 14 px diameter (radius 7) matches the in-tile dot size
    spec'd in the dashboard layout. The header role indicator uses
    radius 9 (18 px diameter) instead — pass ``radius=9``.
    """
    draw = ImageDraw.Draw(image)
    draw.ellipse(
        (cx - radius, cy - radius, cx + radius, cy + radius),
        fill=color,
        outline=p.BG_PRIMARY,  # 1 px black hairline so the dot pops on dark tiles
    )

File: crates/ados-display/src/graphics/status_dot.rs
//! Filled-circle status dot.
//!
//! A small filled disc whose only signal is its color, used for the header role
//! indicator and for in-tile mesh / uplink status. The three-color status
//! semantics map cleanly to success / warning / error, so no shape variation is
//! needed. A 1 px hairline outline in the background color makes the dot pop on
//! a dark tile.
//!
//! The default radius (7, a 14 px disc) matches the in-tile dot size; the
//! header role indicator passes a larger radius and the broadcasting pulse a
//! smaller one.

use embedded_graphics::pixelcolor::Rgb888;

use super::primitives::{fill_circle, Canvas};

/// Default in-tile dot radius — a 14 px disc.
pub const DEFAULT_RADIUS: i32 = 7;

/// Draw a filled disc of `radius` centered at `(cx, cy)` in `color`, with a
/// 1 px `outline` hairline (pass the background color so the dot pops on a dark
/// tile).
pub fn draw_dot(
    canvas: &mut Canvas,
    cx: i32,
    cy: i32,
    color: Rgb888,
    radius: i32,
    outline: Rgb888,
) {
    fill_circle(canvas, cx, cy, radius, color, Some(outline));
}

#[cfg(test)]
mod tests {
    use embedded_graphics::prelude::RgbColor;

    use super::*;

    #[test]
    fn dot_paints_its_center() {
        let mut c = Canvas::new(20, 20, Rgb888::BLACK);
        let color = Rgb888::new(0x22, 0xC5, 0x5E);
        draw_dot(&mut c, 10, 10, color, DEFAULT_RADIUS, Rgb888::BLACK);
        assert_eq!(c.pixel(10, 10), color);
    }

    #[test]
    fn dot_stays_within_its_radius() {
        let mut c = Canvas::new(40, 40, Rgb888::BLACK);
        let color = Rgb888::new(0xEF, 0x44, 0x44);
        draw_dot(&mut c, 20, 20, color, 5, Rgb888::BLACK);
        // A pixel two radii away from center is untouched.
        assert_eq!(c.pixel(20 + 12, 20), Rgb888::BLACK);
    }
}
